mmaprototype: track span config normalization errors

wenyihu6 opened the pull request:
Epic: CRDB-55052 
Release note: none

---
**mmaprototype: extract processRangeMsg from processStoreLeaseholderMsgInternal**

This commit extracts a new function processRangeMsg from
processStoreLeaseholderMsgInternal so that we can use defer() to handle metrics
updates since it does continue in the loop.

---
**mmaprototype: track span config normalization errors**

This change splits the SpanConfigNormalizationError counter into two gauges
(HardError for nil conf, SoftError for usable best-effort conf) and adds a
hasNormalizationError field to rangeState so that errors persist until a new
config arrives.

Note that the gauge is updated in processStoreLeaseholderMsg rather than
rebalanceStores because it already iterates all leaseholder ranges at O(ranges)
cost, whereas rebalanceStores only iterates topK and would require an additional
full iteration.

wenyihu6 committed:
mmaprototype: extract processRangeMsg from processStoreLeaseholderMsgInternal

This commit extracts a new function processRangeMsg from
processStoreLeaseholderMsgInternal so that we can use defer() to handle metrics
updates since it does continue in the loop.

wenyihu6 committed:
mmaprototype: track span config normalization errors

This change splits the SpanConfigNormalizationError counter into two gauges.
SpanConfigNormalizationError counts all normalization errors (both HardError for
nil conf and SoftError for usable best-effort conf), while
SpanConfigNormalizationSoftError counts only soft errors.

Note that the gauge is updated in processStoreLeaseholderMsg rather than
rebalanceStores because it already iterates all leaseholder ranges at O(ranges)
cost, whereas rebalanceStores only iterates topK and would require an additional
full iteration.

tbg reviewed (approved):
> @tbg reviewed 2 files and all commit messages, and made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159872)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6).

___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqM6qA5FTbO7ZZ1Wm6:-OgqM6qA5FTbO7ZZ1Wm7:btbe86r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L54)):*
> ```Go
> }
> 
> // incSpanConfigNormalizationErrorIfNonNil increments the
> ```

stale comment
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqLLXk8RLk4ssnzhJ4:-OgqLLXk8RLk4ssnzhJ5:bij4qwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L185)):*
> ```Go
> 		Measurement: "Range",
> 		Unit:        metric.Unit_COUNT,
> 		LabeledName: "mma.span_config.normalization",
> ```

Apologies for bringing this up again when I'm unsure what your original response to this was on a prior PR: I am a big fan of having fewer metrics we need to check. As is, I believe we need to separately check hard and soft errors. If we had a catch-all errors metric and then one for soft errors, we would only need to track one, and could check the other if there was anything to check. Was there a reason to split the error counters and not have one subsume the other? If this was discussed and this is what was landed on, fine by me.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1577 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqM1EO2BzmubYKKVt9:-OgqM1EO2BzmubYKKVtA:bwcdiay) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1577)):*
> ```Go
> 	// changed.
> 	rs.clearAnalyzedConstraints()
> 	return
> ```

why this return?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: rename counterMetrics to rangeOperationMetrics

This commit renames counterMetrics to rangeOperationMetrics to better describe
its purpose since it now includes gauge metrics for tracking span config
normalization errors.

wenyihu6 reviewed (commented):
These metrics don’t include constraint analysis errors. I think it may be better to panic when normalization succeeds but constraint analysis fails. That only happens if we fail to find a leaseholder here (https://github.com/cockroachdb/cockroach/blob/ee20a2d2c09a1fdd9251351bd014751e1ecb6d36/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go#L765-L777, https://github.com/cockroachdb/cockroach/blob/968fd0f910545e6a736a058061cf3e5cd459656f/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1755-L1757), which should be an unexpected case and already violates mma invariant https://github.com/cockroachdb/cockroach/blob/f4fd4252bcd3de1c56251077999aa08284f1f2d3/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L673. If we agree, I’ll add this to the issue and recover from the panic when it occurs.

If we do want to track constraint analysis errors, constraint analysis happens inside `rebalanceStores` and only applies to the top-K ranges we select as candidates, so it doesn't look at all ranges. That said, if many ranges are failing to rebalance due to constraint analysis errors, that’s more concerning. A gauge isn’t a good fit here though, because it depends on which ranges were selected as candidates, so we can’t reliably tell whether we’ve truly made progress.

> @wenyihu6 made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-:-OgqnY7a-coho8ZhawVT:bykp09l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @sumeerbhola, and @tbg).

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1577 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqM1EO2BzmubYKKVt9:-OgqnEKg9rd1bzU2Ghgy:b-akx91j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1577)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

why this return?
</blockquote></details>

It's because this function now has named return values. I can do `return hasError, hasSoftError` here as well. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqM6qA5FTbO7ZZ1Wm6:-OgqnD7a5W0WEjIqDQhC:btquzjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L54)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

stale comment
</blockquote></details>

Good catch, removed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqLLXk8RLk4ssnzhJ4:-Ogqn-cLCZq4NX4rKYre:b-lhlh9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L185)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Apologies for bringing this up again when I'm unsure what your original response to this was on a prior PR: I am a big fan of having fewer metrics we need to check. As is, I believe we need to separately check hard and soft errors. If we had a catch-all errors metric and then one for soft errors, we would only need to track one, and could check the other if there was anything to check. Was there a reason to split the error counters and not have one subsume the other? If this was discussed and this is what was landed on, fine by me.
</blockquote></details>

Addressed. Agree that this is the better approach. You also mentioned tracking the number of unique span configs that failed with an error. It's a bit more involved, since the range state would need to retain the original user-input `MaybeSpanConf` even when `MaybeSpanConfIsPopulated` is false, so we can accurately track the map across all current ranges. So I opted out for now. 


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> @wenyihu6 made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159872)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @sumeerbhola, and @tbg).

___
*[`-- commits` line 19 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-Ogqs74jDQnLllYZ5r8M:-Ogqs74jDQnLllYZ5r8N:b-46m8fv):*
Reminder to myself - will need to reword this commit message when we merge. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
I agree we shouldn't include constraint analysis errors in these particular error metrics. Having them be in the catch-all panic metric is fine.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 5 files and all commit messages, and made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-:-OgsoLbU6gABfcL41Aj-:bvw17em)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, @tbg, and @wenyihu6).

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1577 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqM1EO2BzmubYKKVt9:-Ogsl4Sj3aVv18ywyufh:b-zhf1jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1577)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

It's because this function now has named return values. I can do `return hasError, hasSoftError` here as well.
</blockquote></details>

I believe the preferred style in the CRDB codebase is for the return values to be listed in the return statement.
And I personally find that easier when reading code.
But given the use of defer to populate the return values, this is fine.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqLLXk8RLk4ssnzhJ4:-OgsnkLF2Cfu6zRsTUhb:bh0gk8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L185)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Addressed. Agree that this is the better approach. You also mentioned tracking the number of unique span configs that failed with an error. It's a bit more involved, since the range state would need to retain the original user-input `MaybeSpanConf` even when `MaybeSpanConfIsPopulated` is false, so we can accurately track the map across all current ranges. So I opted out for now.
</blockquote></details>

I'm fine with the change, but just want to note that having one subsume the other is not best-practice when one has labeled metrics, since then one would want num-error{kind=soft} and num-error{kind=hard} as separate timeseries, and someone wanting the sum would aggregate away the kind. I realize we are not in that world with our internal dbconsole, but we should consider that in the future.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
TFTR! 

bors r+

> @wenyihu6 made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-:-Ogtp-ww4hVSjaE3A0-c:b-4iczgp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DarrylWong, @herkolategan, @sumeerbhola, and @tbg).

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1577 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqM1EO2BzmubYKKVt9:-Ogtovv4AYvwxKi025Ru:bahasxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1577)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I believe the preferred style in the CRDB codebase is for the return values to be listed in the return statement.
And I personally find that easier when reading code.
But given the use of defer to populate the return values, this is fine.
</blockquote></details>

Ack, will keep it as is for now but will keep this in mind in the future. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159872#-OgqLLXk8RLk4ssnzhJ4:-OgtoreY2kKkWtixJNE0:b-ng9o4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f7f8e5910ded3a11bccfba3142e2bb0a5bf5e3a/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L185)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm fine with the change, but just want to note that having one subsume the other is not best-practice when one has labeled metrics, since then one would want num-error{kind=soft} and num-error{kind=hard} as separate timeseries, and someone wanting the sum would aggregate away the kind. I realize we are not in that world with our internal dbconsole, but we should consider that in the future.
</blockquote></details>

Ack, will keep it as is for now. But happy to revisit this in the future. 


<!-- Sent from Reviewable.io -->
